man/res_package.5
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "RES_PACKAGE 1"
.TH RES_PACKAGE 1 "2014-09-13" "Clockwork v2.3.0" "Clockwork Manual"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
res_package \- Clockwork Resource Type for software packages
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The package resource allows administrators to specify what software
should be installed (or not installed) on what hosts.  Package
management is performed through the local system's package manager;
standalone installation scripts (i.e. .bin files) are not supported.
.PP
The following Linux package managers are currently supported:
.IP "\fB\s-1YUM\s0\fR" 4
.IX Item "YUM"
For RPM-based platforms (RedHat, CentOS, SuSE, \s-1OEL,\s0 etc.)
.IP "\fB\s-1APT\s0\fR" 4
.IX Item "APT"
For DPKG-based platforms (Debian, Ubuntu, etc.)
.SH "ATTRIBUTES"
.IX Header "ATTRIBUTES"
.IP "\fBname\fR" 4
.IX Item "name"
The name of the software package, according to the local package
manager.  This may be different based on host architecture; conditional
policy constructs should be used to get the correct name.
.IP "\fBversion\fR" 4
.IX Item "version"
The (optional) version of the package to install.  If not specified,
and the package is not already installed, the latest available version
will be used.  The special value \*(L"any\*(R" forces this default behavior.
.Sp
The special value \*(L"latest\*(R" is will cause Clockwork to keep the installed
package at the latest version, upgrading as needed.
.IP "\fBinstalled\fR" 4
.IX Item "installed"
Whether or not the package should be installed on the system.  Valid
values are \fB\*(L"yes\*(R"\fR and \fB\*(L"no\*(R"\fR.  Default value is \fB\*(L"yes\*(R"\fR.
.SH "EXAMPLES"
.IX Header "EXAMPLES"
.SS "Installation of Packages"
.IX Subsection "Installation of Packages"
Ensure that the apache web server package, httpd, is installed and
up-to-date:
.PP
.Vb 1
\&    package "httpd" { }
.Ve
.PP
Even though the default behavior of the package resource is to install
the software package, explicitly specifying this is considered good
practice:
.PP
.Vb 3
\&    package "httpd" {
\&        installed: "yes"
\&    }
.Ve
.PP
Same as the above, but use the 2.2 version of Apache, not the latest
available version:
.PP
.Vb 4
\&    package "httpd" {
\&        installed: "yes"
\&        version:   "2.2"
\&    }
.Ve
.SS "Removal of Packages"
.IX Subsection "Removal of Packages"
Removing packages is easy; just set \fBinstalled\fR to \fB\*(L"no\*(R"\fR:
.PP
.Vb 3
\&    package "emacs" {
\&        installed: "no"
\&    }
.Ve
.SS "Platform Detection and Package Names"
.IX Subsection "Platform Detection and Package Names"
Not all distributions name their packages the same way.  Using \fIif\fR
constructs in the policy definition makes it easy to accommodate
different naming schemes.
.PP
For this example, assume we want to install \fImypackage\fR on CentOS and
Ubuntu machines in your infrastructure.  The Ubuntu package name is 
\&\fImy-package-core\fR, but the CentOS package name is \fImypackage\fR.
.PP
(We also assume that there are no other platforms to worry about.)
.PP
.Vb 1
\&    package "mypackage" {
\&
\&        # override the package name for Ubuntu
\&        if (lsb.distro.id is "Ubuntu") {
\&            name: "my\-package\-core"
\&        }
\&
\&        installed: "yes"
\&    }
.Ve
.PP
The above resource definition works because the initial package name will
be set to \*(L"mypackage\*(R", and then overidden if the agent is running on Ubuntu.
.SH "CAVEATS"
.IX Header "CAVEATS"
.SS "1. Package Removal and Versioning"
.IX Subsection "1. Package Removal and Versioning"
Version information (set by the \fBversion\fR attribute) is ignored
if the package is set to be removed.
.SH "DEPENDENCIES"
.IX Header "DEPENDENCIES"
None.
.PP
\&\fBNote:\fR
Package dependency resolution (i.e. apache depends on the apr-util
package) is handled solely by the package manager.
.SH "AUTHOR"
.IX Header "AUTHOR"
Clockwork was designed and written by James Hunt.


CHANGELOG
Release Notes for Clockwork


v2.3.0 - 2014-09-11

  * New `cw' multi-call binary to replace cw* tools into a single,
    user-extensible sub-command dispather, a la git.

    The following utilities have therefore been renamed:

       cwcc     ->  cw-cc
       fact     ->  cw-fact
       cwtrust  ->  cw-trust
       cwpol    ->  cw-shell
       cwkey    ->  cw-cert
       cw-run   ->  cw-mesh (and cw-show, cw-ping, etc.)

       cwtool   ->  cw-localsys [internal to cogd]

   * cw-mesh prompting is now done to stderr, for pipelines
   * cw-mesh now ignores nodes that optout by default
     (use --optout to retain the legacy behavior)
   * cw-mesh now honors ~/.cwrc as an adjunct to /etc/clockwork/cw.conf




res_package.c
#include "res_package.h"
#include "report.h"
#include "pack.h"

#define RES_PACKAGE_PACK_PREFIX "res_package::"
#define RES_PACKAGE_PACK_OFFSET 13
/* Pack format for res_package structure:
     L - enforced
     a - name
     a - version
 */
#define RES_PACKAGE_PACK_FORMAT "Laa"

void* res_package_new(const char *key)
{
	struct res_package *rp;

	rp = xmalloc(sizeof(struct res_package));
	list_init(&rp->res);

	rp->enforced = 0;
	rp->different = 0;
	rp->version = NULL;

	if (key) {
		res_package_set(rp, "name", key);
		rp->key = string("res_package:%s", key);
	} else {
		rp->key = NULL;
	}

	return rp;
}

void res_package_free(void *res)
{
	struct res_package *rp = (struct res_package*)(res);
	if (rp) {
		list_del(&rp->res);

		free(rp->name);
		free(rp->version);

		free(rp->key);
	}

	free(rp);
}

const char* res_package_key(const void *res)
{
	const struct res_package *rp = (struct res_package*)(res);
	assert(rp);

	return rp->key;
}

int res_package_norm(void *res) { return 0; }

int res_package_set(void *res, const char *name, const char *value)
{
	struct res_package *rp = (struct res_package*)(res);
	if (strcmp(name, "name") == 0) {
		free(rp->name);
		rp->name = strdup(value);

	} else if (strcmp(name, "version") == 0) {
		free(rp->version);
		rp->version = strdup(value);

	} else if (strcmp(name, "installed") == 0) {
		if (strcmp(value, "no") != 0) {
			rp->enforced ^= RES_PACKAGE_ABSENT;
		} else {
			rp->enforced |= RES_PACKAGE_ABSENT;
		}

	} else {
		return -1;
	}

	return 0;
}

int res_package_stat(void *res, const struct resource_env *env)
{
	struct res_package *rp = (struct res_package*)(res);

	assert(rp);
	assert(env);
	assert(env->package_manager);

	free(rp->installed);
	rp->installed = package_manager_query(env->package_manager, rp->name);

	return 0;
}

struct report* res_package_fixup(void *res, int dryrun, const struct resource_env *env)
{
	struct res_package *rp = (struct res_package*)(res);

	assert(rp);
	assert(env);
	assert(env->package_manager);

	struct report *report = report_new("Package", rp->name);
	char *action;

	if (res_package_enforced(rp, ABSENT)) {
		if (rp->installed) {
			action = string("uninstall package");

			if (dryrun) {
				report_action(report, action, ACTION_SKIPPED);
			} else if (package_manager_remove(env->package_manager, rp->name) == 0) {
				report_action(report, action, ACTION_SUCCEEDED);
			} else {
				report_action(report, action, ACTION_FAILED);
			}
		}

		return report;
	}

	if (!rp->installed) {
		action = (rp->version ? string("install package v%s", rp->version)
		                      : string("install package (latest version)"));

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else if (package_manager_install(env->package_manager, rp->name, rp->version) == 0) {
			report_action(report, action, ACTION_SUCCEEDED);
		} else {
			report_action(report, action, ACTION_FAILED);
		}

		return report;
	}

	if (rp->version && strcmp(rp->installed, rp->version) != 0) {
		action = string("upgrade to v%s", rp->version);

		if (dryrun) {
			report_action(report, action, ACTION_SKIPPED);
		} else if (package_manager_install(env->package_manager, rp->name, rp->version) == 0) {
			report_action(report, action, ACTION_SUCCEEDED);
		} else {
			report_action(report, action, ACTION_FAILED);
		}

		return report;
	}

	return report;
}

int res_package_is_pack(const char *packed)
{
	return strncmp(packed, RES_PACKAGE_PACK_PREFIX, RES_PACKAGE_PACK_OFFSET);
}

char* res_package_pack(const void *res)
{
	const struct res_package *rp = (const struct res_package*)(res);
	assert(rp);

	return pack(RES_PACKAGE_PACK_PREFIX, RES_PACKAGE_PACK_FORMAT,
	            rp->enforced,
	            rp->name, rp->version);
}

void* res_package_unpack(const char *packed)
{
	struct res_package *rp = res_package_new(NULL);

	if (unpack(packed, RES_PACKAGE_PACK_PREFIX, RES_PACKAGE_PACK_FORMAT,
		&rp->enforced,
		&rp->name, &rp->version) != 0) {

		res_package_free(rp);
		return NULL;
	}

	rp->key = string("res_package:%s", rp->name);

	if (rp->version && !*(rp->version)) {
		/* treat "" as NULL */
		free(rp->version);
		rp->version = NULL;
	}

	return rp;
}

